Associate master instance with AWS Elastic IP

madis opened the pull request:
When MASTER_RESERVED_IP is set to elastic IP from AWS, then aws/util.sh will
associate it with master instance and assign it to KUBE_MASTER_IP. If no MASTER_RESERVED_IP
is set, new elastic ip will be requested from amazon. This allows cluster certificates to
be generated for an IP that doesn't change between stopping & starting cluster instances.

The requested elastic ip is not released when kube-down.sh is run. I think it is good
because user could have created DNS records and it would be bad if the IP was removed.
He can reuse it next time through MASTER_RESERVED_IP when setting up cluster again.

madis committed:
Optionally associate master instance with AWS Elastic IP

When MASTER_RESERVED_IP is set to elastic IP from AWS, then aws/util.sh will
associate it with master instance and assign it to KUBE_MASTER_IP. If no MASTER_RESERVED_IP
is set, new elastic ip will be requested from amazon. This allows cluster certificates to
be generated for an IP that doesn't change between stopping & starting cluster instances.

The requested elastic ip is not released when kube-down.sh is run. I think it is good
because user could have created DNS records and it would be bad if the IP was removed.
He can reuse it next time through MASTER_RESERVED_IP when setting up cluster again.

madis commented:
I signed it!

madis commented:
Not sure why travis build failed. Ran the failing command: ```The command "KUBE_RACE="-race" KUBE_COVER="y" KUBE_GOVERALLS_BIN="$HOME/gopath/bin/goveralls" KUBE_TIMEOUT='-timeout 300s' KUBE_COVERPROCS=8 KUBE_TEST_API_VERSIONS=$KUBE_TEST_API_VERSIONS ./hack/test-go.sh -- -p=2" exited with 1.``` locally with success.

roberthbailey commented:
@madis Travis typically runs on slower machines with fewer resources than your development environment and this occasionally exposes race conditions that we don't otherwise see.

/cc @justinsb for AWS changes.

manolitto commented:
Great patch, that solves the changing public IP problem on AWS, thanks! And what is really great for my use case: I can remove the (elastic) public IP from the master instance after setup. Reason why I want to do this: I have a VPN connection to my VPC anyway and I don't want to keep the public IP for security reasons.
What would be even greater: Do you think it is also possible to do this for the minions? So, having MINION_1_RESERVED_IP, MINION_2_RESERVED_IP, ...
Would appreciate this feature very much.

madis commented:
Regarding reserving public ip-s for minions, I'm not sure how well it fits with kubernetes ideology. The way I understand, eventually the minions could come and go, the user should not have to log into (nor having to know addresses of) individual minions. Kubernetes master takes care of what happens inside them. This would make this feature less desirable. Also if the minions come & go, it would be hard to keep track the index of the minion (```MINION_[index]_RESERVED_IP```).

manolitto commented:
Yes, you are right. Addresses of the minions should not be of interest at all. Agreed. Reason, why I suggested elastic IPs for minions is that it is the only way (AFAIK) of being able to remove the public IPs completely right after cluster setup.
The master communicates via the internal IPs and there is no need to log into a certain minion (as you already pointed out). So, why do minions need a public IP at all? Well, without a public IP the minion would not be able to communicate with the internet. Pulling docker images would not work.
The solution for my use case is to use a NAT instance in a public subnet (within the kubernetes VPC) to enable internet access. So, perhaps an even better fit for this case would be a setup feature that enables/disables public IP assignment for minions at all.
Something like ENABLE_MINION_PUBLIC_IP with default value "true". WDYT?

roberthbailey commented on cluster/aws/util.sh:
```diff
@@ -352,6 +352,34 @@ function wait-for-instance-running {
+  local instance_id=$2
+  local elastic_ip_allocation_id=$($AWS_CMD describe-addresses --public-ips $ip_address --output text | cut -f2)
+  $AWS_CMD associate-address --instance-id $master_instance_id --allocation-id $elastic_ip_allocation_id > /dev/null
+  local association_result=$?
```
This is going to go badly, because we set errexit at the top of [./cluster/kube-up.sh](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/kube-up.sh#L23). This means that if the result is != 0 the script will exit right away before you get to check the return value. Elsewhere we set a var and then do `|| var="failed"` or something and then check the value of var after the command finishes.

roberthbailey commented on cluster/aws/util.sh:
```diff
@@ -352,6 +352,34 @@ function wait-for-instance-running {
 }
 
+function allocate-elastic-ip {
+  $AWS_CMD allocate-address --domain vpc --output text | cut -f3
```
Please quote your variables (through the change) as "${AWS_CMD}" with both quotation marks and curly braces.

madis replied:
I'm not sure what you meant. ```"${AWS_CMD}"``` will be treated as string (so not executed). ```"$(AWS_CMD)"``` literally as command.

```bash
#!/usr/bin/env bash
AWS_CMD="ls -la"
echo "${AWS_CMD}" # => ls -la
echo "$(AWS_CMD)" # => ./cli.sh: line 4: AWS_CMD: command not found
```

justinsb replied:
There's a pretty strict bash style (as you've probably noticed!)  I think in this case, we can't quote it, but curly brackets help us get closer to full-adherence to the "style guide".  (I'm not sure if the style guide is actually written down anywhere, but we set options to exit on error, and require variable declaration; also I find it helpful to copy an example from elsewhere in the script when testing e.g. whether a variable is empty)

madis replied:
Is https://google-styleguide.googlecode.com/svn/trunk/shell.xml the proper styleguide to adhere or is it some other?

roberthbailey replied:
Our style guide has been informally set by @filbranden but is probably largely based on the google shell style guide.

filbranden replied:
@madis 

Recommendation (from the style guide you referenced, yes that's the correct one) is to always use `${...}` for variables with more than one character to make that consistent regardless of whether you have some string following the variable that would require the curly braces, so we essentially try to use it everywhere...

If you expect `${AWS_CMD}` to expand to a single word then you should double quote it, but your example of `ls -la` shows that maybe you don't...

In that case, if `${AWS_CMD}` is defined inside your script, you should probably use a bash array instead since that's the more general way to support a variable expanding to multiple arguments that is friendly with whitespace.

Though, if you expect AWS_CMD to be set or exported from the external environment, then it's fair to just use the unquoted `${AWS_CMD}` as you're using it. In that case, though, make sure you handle the way of `${AWS_CMD}` not being set in the environment before this script is called, for instance by defaulting it to something useful, or by checking that it is properly defined before you use it (or at least defaulting it to something like `false` which under `set -o errexit` will trigger a somewhat more useful error when executed instead of taking the next argument and assuming it's a command, in which case you run a risk if that actually looks like a valid command then something unintended will run instead of what you originally expected.)

I hope this addresses your concerns.

manolitto commented:
FYI: I just opened a pull request #7926 which implements the "disable minion public ip" feature.

madis committed:
Properly get return value (considering errexit). Quote variables.

justinsb commented:
So I'm in two minds about this.  It obviously solves a real problem today, but I think long term we will likely have multi-master (and maybe even let the master run on any node?), so this option won't make a lot of sense.

All in all, I think multi-master is a while off (post-V1 I believe), so I think we need something like this.

I also think "leaking" the elastic ip is a problem.

Can we tag elastic ips?  If so, how about we do this:
* We look for an elastic ip tagged for our cluster
* If there is an unused one, we reuse that
* If not, we always allocate one (and tag it!)
* On cluster/kube-down, we delete the elastic ips by default (maybe we need a flag here to prevent deletion, though my guess would be that we wouldn't)

I would prefer that approach because it doesn't leak elastic ips, is a little easier on the end-user, and allows us to have multiple masters in future.

madis commented:
Based on http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html Elastic IP doesn't support tags

justinsb commented:
Bummer on the no-tags on Elastic IPs.   I wonder if there's something we can do that is comparable.  (We could store the IPs e.g. on tags on the VPC, but that feels really hacky).

roberthbailey commented:
If it's any consolation, for GCE we just name the static IP based on the cluster name (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/util.sh#L548) and then delete it when you call kube-down.sh (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/util.sh#L718). 

Would something similar work for AWS?

madis commented:
I amended the last commit based on the feedback. Please let me know if more more changes are needed.

madis commented:
I don seem to have access to the shippable build logs (it remains in *Waiting for build information* indefinitely) and so don't know what caused it to fail. Is there something else to fix in the pull request?

roberthbailey commented:
I'm waiting for @justinsb to give the green light.

justinsb commented:
I think this is important, but I am wary of leaking elastic IPs, and what we do when we go multi-master.  Also, I think for the normal use case (when we assume instances "live forever") we don't have a problem.

How about this for a compromise: if MASTER_RESERVED_IP is not set, we do *not* allocate an elastic IP. If the user sets MASTER_RESERVED_IP, then we use that public IP.  We do not delete it on kube-down, because we didn't create it.

If it is convenient for you, then e.g. if the user sets MASTER_RESERVED_IP=auto then you can allocate an elastic ip.  We still won't treat it as owned, this would just be a convenience.

I think because this is opt-in, this is more acceptable.  Also, by opting-in, the user is saying "I own the public IP", so we don't need to worry about leaking it.

I suspect long-term the answer will be that a cluster of masters self-manages an elastic IP, or that there is an ELB in front of an auto-scaling group of masters.

madis committed:
Add 'auto' option for MASTER_RESERVED_IP. No ElasticIP allocation by default.

Default behaviour when setting up a cluster is using the Amazon-assigned public ip.
It will change between reboots. If MASTER_RESERVED_IP is set to 'auto', new Elastic
IP will be allocated & assigned to master. If MASTER_RESERVED_IP is set to an existing
Elastic IP, it will be used. When something fails, original Amazon-given IP will be used.

madis commented:
I made changes that @justinsb suggested. Now:

1. Default behaviour when setting up a cluster is to use the Amazon-assigned public ip
2. If MASTER_RESERVED_IP is set to 'auto', new Elastic IP will be allocated & assigned to master.
3. If MASTER_RESERVED_IP is set to an existing Elastic IP, it will be used. 

When something fails, original Amazon-given IP will be used.

justinsb commented:
Thanks for making the changes; this LGTM.

madis commented:
Changed *then* to be on the same line as *if* and *elif*.

roberthbailey commented:
I'll take another look in the morning. Hopefully travis will have run by then...

roberthbailey commented:
LGTM.

mikaelh commented:
@madis This feature currently returns "A client error (InvalidInstanceID) occurred when calling the AssociateAddress operation: The pending instance x is not in a valid state for this operation."

AWS seem to require that the (master) instance is in "running" state before an elastic IP can be assigned. An easy fix is to move assign-elastic-ip after wait-for-instance-running has returned.
